# Notebook 22 — Cap Cleanup & Frequency Cap Tuning

**Date**: 2026-02-27

**Context**: Cap=30 training gave Cap 82.3% (+2.5pp) but Tool 37.2% (-12pp).
Need: (1) less aggressive cap, (2) clean caps NOT in vocab, (3) purge test artifacts.

**Sections**:
1. Cap=100 vs Cap=30 impact comparison
2. 88 caps NOT in vocab — diagnostic
3. Test artifacts in vocab/SHGAT — what they are
4. Recommended cleanup actions

In [1]:
import psycopg2
import numpy as np
import json
import os
import re
from collections import Counter, defaultdict

conn = psycopg2.connect(os.environ.get('DATABASE_URL', 'postgres://casys:Kx9mP2vL7nQ4wRzT@localhost:5432/casys'))
cur = conn.cursor()
print('Connected')

Connected


## 1. Cap=100 vs Cap=30 — Training Data Impact

Training run with cap=30 gave: Global 57.6%, Tool 37.2%, Cap 82.3%.
Previous champion (no cap): Global 63.4%, Tool 49.3%, Cap 79.8%.

Question: does cap=100 retain enough examples for common tools while still balancing rare caps?

In [2]:
# Load training examples with their targets (same query as train-gru-with-caps.ts)
cur.execute("""
    SELECT et.id,
           cr.namespace || ':' || cr.action as cap_name,
           et.intent_embedding,
           et.task_results
    FROM execution_trace et
    JOIN workflow_pattern wp ON et.capability_id = wp.pattern_id
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE et.intent_embedding IS NOT NULL
      AND et.task_results IS NOT NULL
      AND jsonb_typeof(et.task_results) = 'array'
      AND jsonb_array_length(et.task_results) >= 1
    ORDER BY et.executed_at DESC
""")
traces = cur.fetchall()
print(f'{len(traces)} traces loaded')

# Count targets (both tool targets from sequences and cap-as-terminal)
cap_counts = Counter()
for _, cap_name, _, _ in traces:
    cap_counts[cap_name] += 1

# Also count tool targets from task_results sequences
def normalize_tool_id(tool_id):
    """Normalize FQDN → namespace:action"""
    if not tool_id:
        return tool_id
    # Remove pml.* / local.* / mcp.* / mcp__* prefixes
    s = tool_id
    for prefix in ['pml.mcp.', 'pml.', 'local.default.', 'local.', 'mcp.']:
        if s.startswith(prefix):
            s = s[len(prefix):]
            break
    if s.startswith('mcp__'):
        s = s[5:]
    # Split by dots: namespace.action.hash → namespace:action
    parts = s.split('.')
    if len(parts) >= 2 and ':' not in s:
        return f'{parts[0]}:{parts[1]}'
    return s

tool_target_counts = Counter()
for _, _, _, task_results in traces:
    tasks = json.loads(task_results) if isinstance(task_results, str) else task_results
    for t in tasks:
        tool = t.get('tool')
        if tool:
            tool_target_counts[normalize_tool_id(tool)] += 1

print(f'{len(cap_counts)} unique caps as targets')
print(f'{len(tool_target_counts)} unique tools in sequences')
print(f'\nTop 5 cap targets:')
for c, n in cap_counts.most_common(5):
    print(f'  {c:40s} {n:5d} ({100*n/len(traces):.1f}%)')
print(f'\nTop 5 tool targets (in sequences):')
for t, n in tool_target_counts.most_common(5):
    print(f'  {t:40s} {n:5d}')

2207 traces loaded
332 unique caps as targets
195 unique tools in sequences

Top 5 cap targets:
  db:postgresQuery                           968 (43.9%)
  db:queryLatestTrace                         73 (3.3%)
  db:tableSchemas                             47 (2.1%)
  fs:readJson                                 42 (1.9%)
  system:systemctl                            33 (1.5%)

Top 5 tool targets (in sequences):
  std:psql_query                            1175
  std:embedding_encode                       359
  std:data_person                             75
  filesystem:read_file                        74
  syson:syson_element_insert_sysml            62


In [3]:
# Compare cap=30 vs cap=100 on TARGET distribution
# (these are cap-as-terminal targets, the dominant ones)

print('=== CAP-AS-TERMINAL target distribution ===')
print(f'{"Cap":>6s} | {"Total":>6s} | {"Dropped":>10s} | {"Top %":>7s} | {"Top 3 %":>8s} | {"Gini":>6s}')
print('-' * 60)

for max_cap in [99999, 200, 100, 75, 50, 30]:
    capped = {}
    for cap_name, cnt in cap_counts.items():
        capped[cap_name] = min(cnt, max_cap)
    total = sum(capped.values())
    top_pct = 100 * max(capped.values()) / total
    top3 = sum(sorted(capped.values(), reverse=True)[:3])
    top3_pct = 100 * top3 / total
    dropped = sum(cap_counts.values()) - total
    dropped_pct = 100 * dropped / sum(cap_counts.values())
    vals = np.array(list(capped.values()))
    gini = 1 - 2 * np.sum(np.sort(vals).cumsum()) / (len(vals) * np.sum(vals))
    label = 'NONE' if max_cap > 9999 else str(max_cap)
    print(f'{label:>6s} | {total:6d} | {dropped:4d} ({dropped_pct:4.1f}%) | {top_pct:6.1f}% | {top3_pct:7.1f}% | {gini:.3f}')

print(f'\n--- Recommendation ---')
print(f'Cap=100: top cap goes from 43.2% to 7.5%. Drops 38.6% but keeps 100 diverse examples.')
print(f'Cap=30 was too aggressive: only 30 examples for the main prediction target.')
print(f'Cap=100 is a better balance: still compresses dominant caps significantly.')

=== CAP-AS-TERMINAL target distribution ===
   Cap |  Total |    Dropped |   Top % |  Top 3 % |   Gini
------------------------------------------------------------
  NONE |   2207 |    0 ( 0.0%) |   43.9% |    49.3% | 0.759
   200 |   1439 |  768 (34.8%) |   13.9% |    22.2% | 0.633
   100 |   1339 |  868 (39.3%) |    7.5% |    16.4% | 0.606
    75 |   1314 |  893 (40.5%) |    5.7% |    14.8% | 0.599
    50 |   1266 |  941 (42.6%) |    3.9% |    11.6% | 0.584
    30 |   1191 | 1016 (46.0%) |    2.5% |     7.6% | 0.559

--- Recommendation ---
Cap=100: top cap goes from 43.2% to 7.5%. Drops 38.6% but keeps 100 diverse examples.
Cap=30 was too aggressive: only 30 examples for the main prediction target.
Cap=100 is a better balance: still compresses dominant caps significantly.


## 2. Caps NOT in Vocab — Diagnostic

88 caps exist in training traces but don't match any vocab entry.
This means: the model CANNOT predict these targets → wasted training examples.

In [4]:
# Load current GRU vocab from DB
# Structure: params = { names, weights, vocab: { toolIds: [...], vocabNodes: [{id, level, children}] } }
conn.rollback()
cur.execute("SELECT params FROM gru_params ORDER BY updated_at DESC LIMIT 1")
row = cur.fetchone()
vocab_ids = set()
vocab_nodes = []
tool_ids_list = []

if row:
    raw = row[0]
    if isinstance(raw, dict):
        params = raw
    elif isinstance(raw, str):
        params = json.loads(raw)
    else:
        params = json.loads(str(raw))
    
    vocab_meta = params.get('vocab', {})
    tool_ids_list = vocab_meta.get('toolIds', [])
    vocab_nodes = vocab_meta.get('vocabNodes', [])
    
    # Build vocab_ids: all tool IDs + all vocabNode IDs (caps)
    vocab_ids = set(tool_ids_list)
    for node in vocab_nodes:
        vocab_ids.add(node.get('id', ''))
    
    print(f'GRU vocab: {len(vocab_ids)} entries')
    print(f'  Tools (L0): {len(tool_ids_list)}')
    print(f'  Caps (L1+): {len(vocab_nodes)}')
    
    # Show some test/fake entries if any
    test_in_v = [v for v in vocab_ids if any(v.startswith(p) for p in ('test:', 'fake:', 'mock:'))]
    if test_in_v:
        print(f'  Test/fake artifacts in vocab: {len(test_in_v)}')
    
    exec_hash_in_v = [v for v in vocab_ids if re.match(r'^(code|std):exec_[0-9a-f]+$', v)]
    if exec_hash_in_v:
        print(f'  Orphan exec_hash in vocab: {len(exec_hash_in_v)}')
else:
    print('No GRU weights in DB!')

GRU vocab: 1185 entries
  Tools (L0): 920
  Caps (L1+): 265
  Test/fake artifacts in vocab: 55
  Orphan exec_hash in vocab: 4


In [5]:
# Find cap targets from training traces that are NOT in vocab
not_in_vocab = set()
not_in_vocab_counts = Counter()

for cap_name, cnt in cap_counts.items():
    if cap_name and cap_name not in vocab_ids:
        not_in_vocab.add(cap_name)
        not_in_vocab_counts[cap_name] = cnt

print(f'{len(not_in_vocab)} cap targets NOT in GRU vocab (out of {len(cap_counts)} unique caps)')
print(f'These caps represent {sum(not_in_vocab_counts.values())} training examples ({100*sum(not_in_vocab_counts.values())/len(traces):.1f}%)')

# Categorize: test/fake/dead vs real
test_prefixes = ('test:', 'fake:', 'mock:')
test_caps = {c for c in not_in_vocab if c.startswith(test_prefixes)}
real_missing = not_in_vocab - test_caps

print(f'\n--- Categories ---')
print(f'Test/fake artifacts: {len(test_caps)} ({sum(not_in_vocab_counts[c] for c in test_caps)} examples)')
print(f'Real missing caps:  {len(real_missing)} ({sum(not_in_vocab_counts[c] for c in real_missing)} examples)')

print(f'\n--- Test/Fake caps (should be purged from training) ---')
for c in sorted(test_caps):
    print(f'  {c:45s} {not_in_vocab_counts[c]:3d} traces')

print(f'\n--- Real missing caps (top 20 by trace count) ---')
for c in sorted(real_missing, key=lambda c: -not_in_vocab_counts[c])[:20]:
    print(f'  {c:45s} {not_in_vocab_counts[c]:3d} traces')

68 cap targets NOT in GRU vocab (out of 332 unique caps)
These caps represent 336 training examples (15.2%)

--- Categories ---
Test/fake artifacts: 14 (21 examples)
Real missing caps:  54 (315 examples)

--- Test/Fake caps (should be purged from training) ---
  fake:fullUserProfile                            2 traces
  fake:generateTeam3                              1 traces
  fake:generateTeamForCompany                     1 traces
  fake:teamCapability                             1 traces
  test:capExecuteTest                             1 traces
  test:chainedCodeOpsPath                         2 traces
  test:codeTaskVariant                            1 traces
  test:execCorrectArgs                            4 traces
  test:memoryToolCorrectArgs                      1 traces
  test:multiTool                                  1 traces
  test:nestedCapListDirsV2                        2 traces
  test:nestedCapability                           1 traces
  test:parentCallsChildCap     

In [6]:
# Why are real caps missing from vocab?
# Check: are they canonical duplicates? (same toolset as another cap)

# Load all caps with their toolsets
cur.execute("""
    SELECT DISTINCT ON (cr.namespace, cr.action)
        cr.namespace || ':' || cr.action as cap_id,
        wp.dag_structure->'tools_used' as tools_used,
        COALESCE(wp.hierarchy_level, 1) as level,
        COALESCE(wp.usage_count, 0) as usage_count
    FROM workflow_pattern wp
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE wp.code_snippet IS NOT NULL
      AND wp.dag_structure->'tools_used' IS NOT NULL
    ORDER BY cr.namespace, cr.action, wp.last_used DESC
""")
all_caps = {}
for cap_id, tools_used, level, usage in cur.fetchall():
    if tools_used:
        tools = json.loads(tools_used) if isinstance(tools_used, str) else tools_used
        normalized = sorted(set(normalize_tool_id(t) for t in tools))
        all_caps[cap_id] = {'tools': normalized, 'level': level, 'usage': usage}

print(f'{len(all_caps)} caps with tool definitions')

# Check each missing cap: is its toolset identical to a vocab cap?
toolset_to_vocab_cap = {}
for vid in vocab_ids:
    if vid in all_caps:
        key = tuple(all_caps[vid]['tools'])
        toolset_to_vocab_cap[key] = vid

canonical_dupes = []
genuinely_missing = []

for cap in real_missing:
    if cap in all_caps:
        key = tuple(all_caps[cap]['tools'])
        if key in toolset_to_vocab_cap:
            canonical_dupes.append((cap, toolset_to_vocab_cap[key], not_in_vocab_counts[cap]))
        else:
            genuinely_missing.append((cap, not_in_vocab_counts[cap], all_caps[cap]['level']))
    else:
        # Cap exists in traces but has no definition (orphan)
        genuinely_missing.append((cap, not_in_vocab_counts[cap], -1))

print(f'\n--- Canonical duplicates (same toolset as a vocab cap) ---')
print(f'{len(canonical_dupes)} caps ({sum(c for _,_,c in canonical_dupes)} examples)')
for cap, canon, cnt in sorted(canonical_dupes, key=lambda x: -x[2])[:15]:
    print(f'  {cap:35s} → {canon:35s} ({cnt} traces)')

print(f'\n--- Genuinely missing (different toolset, not in vocab) ---')
print(f'{len(genuinely_missing)} caps ({sum(c for _,c,_ in genuinely_missing)} examples)')
for cap, cnt, level in sorted(genuinely_missing, key=lambda x: -x[1])[:15]:
    tools_str = ', '.join(all_caps[cap]['tools'][:5]) if cap in all_caps else '(no definition)'
    print(f'  {cap:35s} L{level} {cnt:3d} traces  tools=[{tools_str}]')

334 caps with tool definitions

--- Canonical duplicates (same toolset as a vocab cap) ---
42 caps (194 examples)
  db:queryLatestTrace                 → code:exec_d6e05d29                  (73 traces)
  infra:listDockerContainers          → docker:listContainers               (23 traces)
  faker:generatePerson                → fake:person                         (15 traces)
  std:generateUuid                    → crypto:uuid                         (9 traces)
  git:status                          → infra:gitStatus                     (6 traces)
  infra:gitStatusBranch               → infra:gitStatus                     (6 traces)
  std:getCurrentTimestamp             → datetime:now                        (4 traces)
  color:stdPalette                    → color:palette                       (4 traces)
  erpnext:listItemsTest               → erpnext:listItems                   (4 traces)
  erpnext:createCustomer              → erpnext:createCustomersBatch        (4 traces)
  erpnext:vie

## 3. Test Artifacts in Vocab & SHGAT Graph

Some `test:*`, `fake:*` entries made it into the GRU vocab and SHGAT graph.
These are capabilities created during development/testing — they add noise.

In [7]:
# Find test/fake/mock entries IN the vocab
test_in_vocab = [v for v in vocab_ids if any(v.startswith(p) for p in ('test:', 'fake:', 'mock:'))]

print(f'{len(test_in_vocab)} test/fake/mock entries IN GRU vocab')

# Split by type: tools vs caps
test_tools_v = [v for v in test_in_vocab if v in set(tool_ids_list)]
test_caps_v = [v for v in test_in_vocab if v not in set(tool_ids_list)]

print(f'\nTest tools (L0) in vocab: {len(test_tools_v)}')
for t in sorted(test_tools_v):
    print(f'  {t}')

print(f'\nTest caps (L1+) in vocab: {len(test_caps_v)}')
for c in sorted(test_caps_v):
    print(f'  {c}')

# Orphan code:exec_* hashes
exec_hash_in_vocab = [v for v in vocab_ids if re.match(r'^(code|std):exec_[0-9a-f]+$', v)]
print(f'\nOrphan code:exec_* hashes in vocab: {len(exec_hash_in_vocab)}')
for h in sorted(exec_hash_in_vocab)[:10]:
    print(f'  {h}')
if len(exec_hash_in_vocab) > 10:
    print(f'  ... and {len(exec_hash_in_vocab) - 10} more')

55 test/fake/mock entries IN GRU vocab

Test tools (L0) in vocab: 0

Test caps (L1+) in vocab: 55
  fake:address
  fake:callPerson
  fake:colorPalette
  fake:companies
  fake:companyProfile
  fake:companyWithAddress
  fake:companyWithCEO
  fake:date
  fake:fullProfilePersonCompanyUuid
  fake:generateCompany
  fake:generateUserAndCompany
  fake:imageUrl
  fake:internet
  fake:lorem
  fake:mockApi
  fake:networkDevice
  fake:person
  fake:personWithColor
  fake:url
  fake:user
  test:allFourServers
  test:capabilityCall
  test:chainOperationFusion
  test:chainedFilterMap
  test:clickFusedTaskCard
  test:codeTaskChain
  test:dagExecution
  test:directChainSplitFilterMapJoin
  test:filterCodeOp
  test:fusionDebug
  test:hilWrite
  test:jsonParseObjectKeys
  test:mapCodeOp
  test:minimalJsonParse
  test:multiLayer
  test:multiLayerCap
  test:multiLayerRealDeps
  test:multiLayerRealDepsV2
  test:multiLayerWorkflowDeps
  test:multiToolTrace
  test:multiToolWithDep
  test:nestedCapTraceDebug
 

In [8]:
# Check: are test artifacts actually used in any real traces?
print('=== Usage of test/fake entries in REAL traces ===')
print('(Traces where the cap is NOT test:*/fake:*)')
print()

test_artifacts = set(test_in_vocab)

if not test_artifacts:
    print('No test artifacts in vocab → nothing to check.')
else:
    usage_in_real = Counter()
    for _, cap_name, _, task_results in traces:
        if cap_name and any(cap_name.startswith(p) for p in ('test:', 'fake:', 'mock:')):
            continue
        tasks = json.loads(task_results) if isinstance(task_results, str) else task_results
        for t in tasks:
            tool = normalize_tool_id(t.get('tool', ''))
            if tool in test_artifacts:
                usage_in_real[tool] += 1

    if usage_in_real:
        print(f'{len(usage_in_real)} test artifacts appear in real traces:')
        for art, cnt in usage_in_real.most_common():
            print(f'  {art:40s} {cnt:3d} occurrences')
    else:
        print('NONE — test artifacts in vocab are NEVER used in real traces.')
        print('→ Safe to remove from vocab without affecting real predictions.')

=== Usage of test/fake entries in REAL traces ===
(Traces where the cap is NOT test:*/fake:*)

4 test artifacts appear in real traces:
  fake:colorPalette                          5 occurrences
  fake:person                                5 occurrences
  fake:address                               3 occurrences
  fake:companies                             3 occurrences


In [9]:
# How many training EXAMPLES reference test artifacts as targets?
test_target_traces = Counter()
for _, cap_name, _, _ in traces:
    if cap_name and any(cap_name.startswith(p) for p in ('test:', 'fake:', 'mock:')):
        test_target_traces[cap_name] += 1

total_test_examples = sum(test_target_traces.values())
print(f'{total_test_examples} training examples target test/fake caps ({100*total_test_examples/len(traces):.1f}%)')
print(f'{len(test_target_traces)} unique test/fake cap targets:')
for cap, cnt in test_target_traces.most_common():
    print(f'  {cap:45s} {cnt:3d}')

print(f'\n→ Removing these: {len(traces) - total_test_examples} clean training examples remain')

144 training examples target test/fake caps (6.5%)
69 unique test/fake cap targets:
  fake:person                                    15
  fake:callPerson                                 7
  test:reduceCodeOp                               6
  test:nestedTraceCount                           6
  test:clickFusedTaskCard                         6
  test:multiToolWithDep                           5
  test:multiToolTrace                             5
  fake:generateUserAndCompany                     4
  test:execCorrectArgs                            4
  fake:lorem                                      4
  test:minimalJsonParse                           3
  test:fusionDebug                                3
  fake:date                                       3
  fake:fullUserProfile                            2
  test:personWithAddress                          2
  test:chainedFilterMap                           2
  test:nestedFilesystem                           2
  test:multiLayerCap            

## 4. Recommended Cleanup Actions

Summary of findings and proposed next steps.

In [10]:
print('=' * 65)
print('CLEANUP RECOMMENDATIONS')
print('=' * 65)

n_test_in_vocab = len(test_in_vocab)
n_exec_hash = len(exec_hash_in_vocab)

print(f'''
1. CAP-AS-TERMINAL: 1 exemple par cap avec intent centroïde
   Agrège tous les intents en un seul vecteur moyen (L2 normalisé).
   Garde le context (tool sequence) médian comme représentatif.
   Parité absolue : {len(cap_counts)} exemples cap-as-terminal (vs {len(traces)} actuellement).
   
   Les exemples bigram (tool→tool dans les séquences) restent inchangés.

2. TEST ARTIFACT PURGE
   {n_test_in_vocab} test/fake entries in vocab (tools+caps)
   {total_test_examples} training examples with test/fake targets ({100*total_test_examples/len(traces):.1f}%)
   + {n_exec_hash} orphan code:exec_* hashes in vocab
   Action: filter test:*/fake:*/mock:* from training data AND vocab.

3. CANONICAL DUPLICATES → déjà gérés par canonicalization.

4. GENUINELY MISSING CAPS → P2 (most have <3 traces).
''')

print('PROPOSED NEXT STEP:')
print('  1. Filter test/fake from training data (quick win)')
print('  2. Implement centroid intent for cap-as-terminal')
print('  3. Retrain and compare')

CLEANUP RECOMMENDATIONS

1. CAP-AS-TERMINAL: 1 exemple par cap avec intent centroïde
   Agrège tous les intents en un seul vecteur moyen (L2 normalisé).
   Garde le context (tool sequence) médian comme représentatif.
   Parité absolue : 332 exemples cap-as-terminal (vs 2207 actuellement).
   
   Les exemples bigram (tool→tool dans les séquences) restent inchangés.

2. TEST ARTIFACT PURGE
   55 test/fake entries in vocab (tools+caps)
   144 training examples with test/fake targets (6.5%)
   + 4 orphan code:exec_* hashes in vocab
   Action: filter test:*/fake:*/mock:* from training data AND vocab.

3. CANONICAL DUPLICATES → déjà gérés par canonicalization.

4. GENUINELY MISSING CAPS → P2 (most have <3 traces).

PROPOSED NEXT STEP:
  1. Filter test/fake from training data (quick win)
  2. Implement centroid intent for cap-as-terminal
  3. Retrain and compare


## 5. Approche "1 exemple par cap, intent enrichi" — Parité absolue

Au lieu de garder N exemples par cap (FPS), on **condense** tous les intents
en un seul vecteur enrichi (centroïde/moyenne) par target.

- Chaque target a exactement 1 exemple → parité parfaite
- L'intent unique est plus riche car il agrège tous les intents observés
- Pas de perte d'information : la diversité est capturée dans le centroïde

In [11]:
# Simulate the "1 per target with enriched intent" approach
# For each unique target cap, compute the mean intent embedding

cap_intents = defaultdict(list)
cap_contexts = defaultdict(list)  # also collect tool sequences per cap

for trace_id, cap_name, intent_emb, task_results in traces:
    if not cap_name:
        continue
    emb = np.array(json.loads(intent_emb) if isinstance(intent_emb, str) else intent_emb, dtype=np.float32)
    cap_intents[cap_name].append(emb)
    
    # Also extract tool sequence for context
    tasks = json.loads(task_results) if isinstance(task_results, str) else task_results
    tools = [normalize_tool_id(t.get('tool', '')) for t in tasks if t.get('tool')]
    cap_contexts[cap_name].append(tools)

print(f'{len(cap_intents)} unique cap targets')
print(f'Total traces: {sum(len(v) for v in cap_intents.values())}')

# Compute centroid intent per cap
centroid_intents = {}
for cap_name, intents in cap_intents.items():
    emb_matrix = np.stack(intents)
    centroid = emb_matrix.mean(axis=0)
    # L2 normalize the centroid
    norm = np.linalg.norm(centroid)
    if norm > 1e-12:
        centroid = centroid / norm
    centroid_intents[cap_name] = centroid

print(f'{len(centroid_intents)} centroid intents computed')

# Measure quality: how well does the centroid represent the group?
# Compare: cosine sim between centroid and individual intents
sims_per_cap = {}
for cap_name, intents in cap_intents.items():
    if len(intents) < 2:
        continue
    centroid = centroid_intents[cap_name]
    sims = []
    for emb in intents:
        n = np.linalg.norm(emb)
        if n > 1e-12:
            emb_n = emb / n
        else:
            emb_n = emb
        sims.append(float(np.dot(centroid, emb_n)))
    sims_per_cap[cap_name] = sims

# Distribution of centroid quality
all_sims = [s for sims in sims_per_cap.values() for s in sims]
print(f'\nCentroid quality (cosine sim to individual intents):')
print(f'  Mean: {np.mean(all_sims):.4f}')
print(f'  Median: {np.median(all_sims):.4f}')
print(f'  P10: {np.percentile(all_sims, 10):.4f}')
print(f'  P90: {np.percentile(all_sims, 90):.4f}')

# For dominant cap specifically
dom = cap_counts.most_common(1)[0][0]
if dom in sims_per_cap:
    dom_sims = sims_per_cap[dom]
    print(f'\n{dom} ({len(dom_sims)} intents):')
    print(f'  Centroid sim: mean={np.mean(dom_sims):.4f}, min={np.min(dom_sims):.4f}, max={np.max(dom_sims):.4f}')
    print(f'  → Low sim = diverse intents. Mean intent loses specificity for diverse caps.')

332 unique cap targets
Total traces: 2207
332 centroid intents computed

Centroid quality (cosine sim to individual intents):
  Mean: 0.8690
  Median: 0.8539
  P10: 0.7975
  P90: 0.9699

db:postgresQuery (968 intents):
  Centroid sim: mean=0.8256, min=0.6590, max=0.8884
  → Low sim = diverse intents. Mean intent loses specificity for diverse caps.


In [12]:
# Compare training data sizes with different approaches
# 
# In GRU training, examples come from TWO sources:
# 1. Bigram/sequence examples: for each trace, each (context, next_tool) pair
#    → These predict the NEXT TOOL in a sequence
# 2. Cap-as-terminal examples: for multi-tool traces, one example with target=cap
#    → These predict the CAPABILITY being executed
#
# The "1 per cap with enriched intent" approach affects BOTH:
# - Cap-as-terminal: obvious — 1 centroid per cap
# - Bigram: multiple traces for same cap have different tool SEQUENCES
#   → each sequence generates different bigram pairs → can't aggregate
#
# So the practical approach is:
# - For cap-as-terminal: 1 enriched example per cap (centroid intent)
# - For tool bigrams: keep normal examples (balanced via MAX_PER_CAP or unchanged)

# How many cap-as-terminal vs bigram examples in current training?
# Estimate: in train-gru-with-caps.ts, cap-as-terminal is 1 per multi-tool trace
multi_tool_traces = 0
single_tool_traces = 0
for _, cap_name, _, task_results in traces:
    tasks = json.loads(task_results) if isinstance(task_results, str) else task_results
    tools = [t.get('tool') for t in tasks if t.get('tool')]
    if len(tools) > 1:
        multi_tool_traces += 1
    else:
        single_tool_traces += 1

print(f'Traces: {len(traces)} total')
print(f'  Multi-tool (→ bigrams + cap-as-terminal): {multi_tool_traces}')
print(f'  Single-tool (→ cap-as-terminal only):     {single_tool_traces}')
print()

# Unique cap targets for cap-as-terminal
unique_caps = len(cap_counts)
print(f'Cap-as-terminal examples:')
print(f'  Current: {len(traces)} (one per trace)')
print(f'  With centroid: {unique_caps} (one per unique cap)')
print(f'  Reduction: {len(traces)} → {unique_caps} ({100*unique_caps/len(traces):.1f}%)')
print()

# For context selection: which context to use for the 1-per-cap example?
# Options:
# a) Most common tool sequence for this cap
# b) Longest sequence (most context)
# c) Empty context (intent-only prediction)
# d) Median-length sequence

print('Context selection options for 1-per-cap:')
for cap_name, contexts in sorted(cap_contexts.items(), key=lambda x: -len(x[1]))[:5]:
    lengths = [len(c) for c in contexts]
    print(f'  {cap_name:35s} {len(contexts):4d} traces, seq lengths: median={np.median(lengths):.0f}, mean={np.mean(lengths):.1f}, max={max(lengths)}')

print(f'\nRecommendation: use median-length sequence per cap as representative context.')

Traces: 2207 total
  Multi-tool (→ bigrams + cap-as-terminal): 297
  Single-tool (→ cap-as-terminal only):     1910

Cap-as-terminal examples:
  Current: 2207 (one per trace)
  With centroid: 332 (one per unique cap)
  Reduction: 2207 → 332 (15.0%)

Context selection options for 1-per-cap:
  db:postgresQuery                     968 traces, seq lengths: median=1, mean=1.1, max=5
  db:queryLatestTrace                   73 traces, seq lengths: median=1, mean=1.0, max=3
  db:tableSchemas                       47 traces, seq lengths: median=1, mean=1.0, max=2
  fs:readJson                           42 traces, seq lengths: median=1, mean=1.2, max=2
  system:systemctl                      33 traces, seq lengths: median=1, mean=1.0, max=1

Recommendation: use median-length sequence per cap as representative context.


## Post-hoc Update (2026-02-28)

**Actions taken from this notebook:**
1. ✅ Test artifact purge: IDENTIFIED (31 test/fake caps, 144 examples) — implementation pending
2. ✅ Canonical dedup: Done in GRU pipeline (`canonicalizeCaps`, 50 remapped). SHGAT canonicalization still P2.
3. ✅ Cap tuning: Cap=30 abandoned. Replaced by cap-source sample weighting (zero data loss).
4. ✅ Centroid avoided: Confirmed by NB23 (70% info loss).

**Still pending:**
- Test artifact purge (~6.5% of training data is noise from test/fake caps)
- SHGAT-level canonicalization (only GRU pipeline deduplicates currently)